### DISCUSSION TITLE
Filtering in getGlobalScope

### DISCUSSION THREAD

**Discussion by antoinelilly** [eclipse-langium/langium#1311] [Q&A] [answered]
Hello ! 
I am trying to override getGlobalScope. My goal is to filter available nodes by documents URI so my first step was to try and restrict getGlobalScope to only scoping through the current document and one more. My goal is to add more URIs later but I don't seem to be able to filter correctly. I have 2 files named test.mydsl and test2.mydsl in the same folder. To make sure I entered my second URI correctly, I copied the string I found in current_uri and changed test.mydsl by test2.mydsl.

Here is what I wrote :

```
protected override getGlobalScope(referenceType: string, context: ReferenceInfo): Scope {
        //Step 1: Retrieve current document URI
        const targetDoc = getDocument(context.container);
        let modeldoc = targetDoc.parseResult.value as Model
        let current_uri = targetDoc.textDocument.uri

        //Step 2: Add it to a set of string and pass it as argument in this.indexManager.allElements
        let other_uri = [another uri I entered manually]
        let uris:Set<string> = new Set([current_uri, other_uri])
        return this.globalScopeCache.get(referenceType, () => new MapScope(this.indexManager.allElements(referenceType, uris)));
```

My problem is that despite this filter, I still can't access the nodes of my other documents. I can access them though if I remove the uris argument, and basically call super.getGlobalScope. Is there something I misunderstand? Do I misuse my URIs? 

Also I have been struggling a lot with this issue, I would appreciate any help on that topic.

**Comment by msujew** [comment, marked as answer]
Hey @antoinelilly,

I haven't tried out your code yet, but I'm fairly sure the way you are performing the caching in the last line leads to a few issues. Basically, the first time any scoping happens for a given `referenceType`, the `new MapScope(this.indexManager.allElements(referenceType, uris))` value will be cached for it. Afterwards, even for different documents (which will result in a different set of `uris`), it will reuse the same scope that was stored the first time. You cannot use the global scope cache if your global scope is on a per-document basis.

Does it work as expected if you remove the caching in the last step and instead return the `MapScope` directly?

**Reply by antoinelilly** [reply]
Yes ! It does ! It wasn't clear to me how the cache worked. Thank you for clarifying and for your quick answer.
